### PR TITLE
Stability - s3_lifecycle

### PR DESCRIPTION
**Pull request by tremble**
##### SUMMARY

Attempt to get the s3_lifecycle tests back in service

- Use AWSRetry
- Add waiters
- Fix "changed" behaviour when dealing with dates

##### ISSUE TYPE

- Feature Pull Request
- Tests Pull Request

##### COMPONENT NAME

s3_lifecycle

##### ADDITIONAL INFORMATION

**DEPENDS-ON:** https://github.com/ansible-collections/amazon.aws/pull/271

### CODE REVIEW

File: plugins/modules/s3_lifecycle.py [hunk 1]
```diff
@@ -10,105 +10,110 @@
...
+      - If C(enabled), the rule is currently being applied.
+      - If C(disabled), the rule is not currently being applied.
     default: enabled
     choices: [ 'enabled', 'disabled' ]
     type: str
   storage_class:
     description:
-      - "The storage class to transition to."
-      - "The 'standard_ia' class is only being available from Ansible version 2.2."
+      - The storage class to transition to.
```
Comment (alinabuzachis) [inline plugins/modules/s3_lifecycle.py:96]: Could you also update the available storage classes? https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Bucket.upload_fileobj

File: plugins/modules/s3_lifecycle.py [hunk 2]
```diff
@@ -10,105 +10,110 @@
...
+      - If no prefix is specified, the rule will apply to the whole bucket.
     type: str
   purge_transitions:
     description:
-      - >
-        "Whether to replace all the current transition(s) with the new transition(s). When false, the provided transition(s)
-        will be added, replacing transitions with the same storage_class. When true, existing transitions will be removed and
-        replaced with the new transition(s)
+      - Whether to replace all the current transition(s) with the new transition(s).
+      - When false, the provided transition(s) will be added, replacing transitions
```
Comment (tremble) [inline plugins/modules/s3_lifecycle.py:44]:
```suggestion
      - When C(false), the provided transition(s) will be added, replacing transitions
```

File: plugins/modules/s3_lifecycle.py (hunk 1 above)
Comment (tremble) [reply plugins/modules/s3_lifecycle.py:96]: Which class is missing?  I only see the same 5 as listed in choices

Comment (alinabuzachis) [reply plugins/modules/s3_lifecycle.py:96]: My bad, sorry! You're right.

**Review by alinabuzachis — APPROVED** [review]

File: plugins/modules/s3_lifecycle.py (hunk 1 above)
Comment (tremble) [reply plugins/modules/s3_lifecycle.py:96]: Should anyone else be looking: https://github.com/boto/botocore/blob/develop/botocore/data/s3/2006-03-01/service-2.json#L9233  (TransitionStorageClass) will tell you what is supported by botocore

**Comment by tremble** [comment]
Tests passing and approval in place.  Merging

**Pull request merged**